### PR TITLE
IGNITE-23572 Nodes that returned back after majority loss could be returned back to stable

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-23572

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java [hunk 1]
```diff
@@ -230,4 +230,64 @@ void testNoHaRecoveryForScZone() throws InterruptedException {
...
+
+        startNode(4);
+
+        createHaZoneWithTable();
+
+        IgniteImpl node = igniteImpl(0);
+
+        Set<String> allNodes = runningNodes().map(Ignite::name).collect(Collectors.toUnmodifiableSet());
+
+        stopNodes(3);
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java]: done

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/ZoneRebalanceUtil.java [hunk 2]
```diff
@@ -155,7 +158,33 @@ public static CompletableFuture<Void> updatePendingAssignmentsKeys(
 
         ByteArray partAssignmentsStableKey = stablePartAssignmentsKey(zonePartitionId);
 
-        Set<Assignment> partAssignments = calculateAssignmentForPartition(dataNodes, partNum, replicas);
+        Set<Assignment> calculatedAssignments = calculateAssignmentForPartition(dataNodes, partNum, replicas);
+
+        Set<Assignment> partAssignments;
+
+        if (consistencyMode == ConsistencyMode.HIGH_AVAILABILITY) {
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/ZoneRebalanceUtil.java:165]: improved javadocs

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java [hunk 3]
```diff
@@ -230,4 +230,64 @@ void testNoHaRecoveryForScZone() throws InterruptedException {
...
+
+        assertRecoveryKeyIsEmpty(node);
+
+        stopNodes(1, 2, 3, 4);
+
+        waitAndAssertRecoveryKeyIsNotEmpty(node);
+
+        assertRecoveryRequestForHaZoneTable(node);
+
+        waitAndAssertStableAssignmentsOfPartitionEqualTo(node, HA_TABLE_NAME, Set.of(0, 1), Set.of(node.name()));
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java]: done

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/GroupUpdateRequest.java [hunk 4]
```diff
@@ -337,7 +336,43 @@ private static CompletableFuture<Integer> partitionUpdate(
...
+                    LOG.info(
+                            "Force update metastore pending partitions key [key={}, partition={}, table={}, newVal={}]",
+                            pendingChangeTriggerKey(partId).toString(),
+                            partId.partitionId(),
+                            partId.tableId(),
+                            nextAssignment
+                    );
+
+                    break;
+                case ASSIGNMENT_NOT_UPDATED:
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/GroupUpdateRequest.java]: removede

**Review by kgusakov — APPROVED** [review]

**Pull request merged**